game-core/src/main/java/games/strategy/triplea/ui/menubar/TripleAMenuBar.java

```diff
@@ -39,7 +38,7 @@ public TripleAMenuBar(final TripleAFrame frame) {
 
     final Optional<InGameLobbyWatcherWrapper> watcher = frame.getInGameLobbyWatcher();
     watcher.filter(InGameLobbyWatcherWrapper::isActive).ifPresent(this::createLobbyMenu);
-    if (!(frame.getGame().getMessenger() instanceof HeadlessServerMessenger)) {
+    if (frame.getGame().getMessengers().isConnected()) {
```

Comment:
To keep in mind, refactorings are not always going to be globally better, sometimes a step back is required for two steps forwards.

Regardless, few benefits:
- removal of the class cast. Generally class cast is considered a class design smell. Further benefits are maybe subjective, I rate removing class casts highly as it adds compile-time
- HeadlessServerMessenger is potentially to be deleted:
  - HeadlessServerMessenger is a no-op API, generally a code-smell.
  - The abstraction to make network transparent is good, but to remove reflective network calls it is better that we migrate the concern. While working through the interactions, it looks like in at least sound code we really suffer from calling 'localhost' with reflective calls for what could be just plain local calls. To remove the reflective code, lopping off HeadlessServerMessenger and moving the 'network-transparent' abstraction seems to be a next step.
 - headlessServerMessenger is badly named, it has nothing to do with bots, it's just a no-op messenger for local games. It has nothing to do with bots and the bot flag is always false when using HeadlessServerMessenger. Since we are burying no-op code, we complicate caller code, the no-op can be rolled up for simpler code and help us disentangle local vs network, which helps us disentangle the reflective network calls

Reply:
>  it's not the exact equivalent.

I think they potentially are, at least in intent. The messengers exist for any network game, and should always exist for bot games. If the network is not connected, that is the same as not having messengers. The HeadlessServerMessenger returns 'false' all the time. 

The other messengers:

`AbstractServerMessengers`:
```
  @Override
  public boolean isConnected() {
    return !shutdown;
  }
```
And client messenger:
```
  @Override
  public boolean isConnected() {
    return socketChannel.isConnected();
  }
```

Checking more in client code, if we fail to connect, there are a various exceptions and likely we will not be rendering a `network` menu anyways. EG:
```
        } catch (final ConnectException e) {
          throw new RuntimeException(String.format("Could not connect host: %s, port: %s, name: %s, mac: %s",
              host, port, name, mac), e);
        }
```

I think the `isConnected()` check is the API that this code probably should have used to begin with, I'm not sure which came first, the `isConnected` method or class cast. The answer to that question is immaterial as the class cast can and should be replaced.

Reply:
Thanks for the in-depth explanation 👍